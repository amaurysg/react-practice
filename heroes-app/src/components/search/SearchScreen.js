import React, { useMemo } from "react";
//Import query-string ... also download package npm
import queryString from "query-string";
import { useHistory, useLocation } from "react-router";
import HeroCard from "../heroes/HeroCard";
import useForm from "../hooks/useForm";
import getHeroByName from "../../selectors/getHeroByName";

const SearchScreen = () => {
  const history = useHistory();
  //import useLocation to search location.search //{}
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  //Here implement useForm with your values
  const [values, handleInputChange] = useForm({
    //searchText is name in input
    searchText: q,
  });

  const { searchText } = values;
  const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

  // console.log(location.search);

  //desestructure {q} /query/

  //Here called data from heroes (import above)

  //desestructure

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search</h4>
          <form action="" onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find your hero"
              autoComplete="off"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Go...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === "" && <div className="alert alert-info">Search Hero</div>}
          {q !== "" && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">{`There is not hero ${q}`}</div>
          )}

          {heroesFiltered.map((hero) => {
            return <HeroCard key={hero.id} {...hero} />;
          })}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
