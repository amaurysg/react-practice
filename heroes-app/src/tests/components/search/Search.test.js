import { mount } from "enzyme"
import { MemoryRouter, Route, useLocation } from "react-router-dom"
import SearchScreen from "../../../components/search/SearchScreen"

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn()
}));

describe('Test in SearchScreen', () => {

  test('Debe mostrar correctamente con valores por defecto', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search Hero')

  })

  test('Debe mostrar batman y el input con el valor del queryString', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper.find('input').prop('value')).toBe('batman')
    //se ejecuto un nuevo snapshot al aparecer la card de batman
    expect(wrapper).toMatchSnapshot()

  })

  test('debe de mostrar error si no se encuentra el Hero!', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper.find('.alert-danger').exists()).toBe(true)

  })

  test('debe llamar el push del history', () => {
    const hero = 'superman';
    const query = `q=${hero}`;
    const history = {
      push: jest.fn()
    }

    useLocation.mockReturnValueOnce({ search: query });

    const wrapper = mount(
      <MemoryRouter>
        <SearchScreen history={history} />
      </MemoryRouter>
    );

    wrapper.find('form').prop('onSubmit')({ preventDefault() { } });

    expect(wrapper).toMatchSnapshot();

    // expect(wrapper.find('input').prop('value')).toBe(hero);
    // expect(history.push).toHaveBeenCalledWith(hero)
    expect(history.push).toHaveBeenCalled();
    expect(history.push).toHaveBeenCalledTimes();





  })





})
