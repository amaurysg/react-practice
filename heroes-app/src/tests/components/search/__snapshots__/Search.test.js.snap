// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in SearchScreen Debe mostrar batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        action=""
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Go...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <a
        className="my-card animate__animated animate__fadeIn"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`Test in SearchScreen Debe mostrar correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        action=""
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Go...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search Hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in SearchScreen debe llamar el push del history 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        action=""
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Go...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search Hero
      </div>
    </div>
  </div>
</div>
`;
